x/stake/keeper.go

```diff
@@ -33,319 +33,315 @@ func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, codespace sdk.
 	return keeper
 }
 
-// get the current in-block validator operation counter
-func (k Keeper) getCounter(ctx sdk.Context) int16 {
-	store := ctx.KVStore(k.storeKey)
-	b := store.Get(CounterKey)
-	if b == nil {
-		return 0
-	}
-	var counter int16
-	err := k.cdc.UnmarshalBinary(b, &counter)
-	if err != nil {
-		panic(err)
-	}
-	return counter
-}
-
-// set the current in-block validator operation counter
-func (k Keeper) setCounter(ctx sdk.Context, counter int16) {
-	store := ctx.KVStore(k.storeKey)
-	bz, err := k.cdc.MarshalBinary(counter)
-	if err != nil {
-		panic(err)
-	}
-	store.Set(CounterKey, bz)
-}
-
 //_________________________________________________________________________
 
-// get a single candidate
-func (k Keeper) GetCandidate(ctx sdk.Context, addr sdk.Address) (candidate Candidate, found bool) {
+// get a single validator
+func (k Keeper) GetValidator(ctx sdk.Context, addr sdk.Address) (validator Validator, found bool) {
 	store := ctx.KVStore(k.storeKey)
-	b := store.Get(GetCandidateKey(addr))
+	b := store.Get(GetValidatorKey(addr))
 	if b == nil {
-		return candidate, false
+		return validator, false
 	}
-	err := k.cdc.UnmarshalBinary(b, &candidate)
-	if err != nil {
-		panic(err)
-	}
-	return candidate, true
+	k.cdc.MustUnmarshalBinary(b, &validator)
+	return validator, true
 }
 
-// Get the set of all candidates, retrieve a maxRetrieve number of records
-func (k Keeper) GetCandidates(ctx sdk.Context, maxRetrieve int16) (candidates Candidates) {
+// Get the set of all validators, retrieve a maxRetrieve number of records
+func (k Keeper) GetValidators(ctx sdk.Context, maxRetrieve int16) (validators Validators) {
 	store := ctx.KVStore(k.storeKey)
-	iterator := store.Iterator(subspace(CandidatesKey))
+	iterator := store.SubspaceIterator(ValidatorsKey)
 
-	candidates = make([]Candidate, maxRetrieve)
+	validators = make([]Validator, maxRetrieve)
 	i := 0
 	for ; ; i++ {
 		if !iterator.Valid() || i > int(maxRetrieve-1) {
 			iterator.Close()
 			break
 		}
 		bz := iterator.Value()
-		var candidate Candidate
-		err := k.cdc.UnmarshalBinary(bz, &candidate)
-		if err != nil {
-			panic(err)
-		}
-		candidates[i] = candidate
+		var validator Validator
+		k.cdc.MustUnmarshalBinary(bz, &validator)
+		validators[i] = validator
 		iterator.Next()
 	}
-	return candidates[:i] // trim
+	return validators[:i] // trim
```

Comment:
that's incorrect - always more efficient to predefine the size over using append... if there was an extreme discrepancy I don't think it would be a problem, the memory would be allocated at the beginning (remember max of a int64 ~32K).

Reply:
OK thanks for the clarification.

Reply:
Why do we need maxRetrieve  here ? Can't we just return all of them  ?